libs/types-common/src/Data/Domain.hs

```diff
@@ -43,12 +74,52 @@ instance FromJSON Domain where
 
 instance Arbitrary Domain where
   arbitrary =
-    either (error "arbitrary @Domain") id . mkDomain
-      <$> elements
-        [ "example.com",
-          "beispiel.com"
-          -- unicode domains are not supported, sadly:
-          -- "例.com",
-          -- "مثال.com",
-          -- "dæmi.com"
-        ]
+    either (error . ("arbitrary @Domain: " <>)) id . mkDomain . getDomainText <$> arbitrary
+
+-- | only for QuickCheck
+newtype DomainText
+  = DomainText {getDomainText :: Text}
+  deriving (Eq, Show)
+
+instance Arbitrary DomainText where
+  arbitrary = DomainText <$> domain
+    where
+      -- <domain> ::= <label> | <domain> "." <label>
+      domain =
+        QC.oneof
+          [ label,
+            conc [domain, pure ".", label]
+          ]
+          `QC.suchThat` ((<= 255) . Text.length)
+      -- <label> ::= <let-dig> [ [ <ldh-str> ] <let-dig> ]
+      label =
+        conc [letDig, opt (conc [opt ldhStr, letDig])]
```

Comment:
We had this cool parser combinator framework at UU that would  generate a parser, pretty printer, and quickcheck generator for each  AST that you defined.  I miss it now.  (As this is just the same as the parser code, but in a different applicative :P; wonder if we can abuse that fact somehow (Free applicative?) )

Ohwell not that important; just an observation

Reply:
I remember seeing something like that and yes, it's nice.

The parser doesn't adhere to the grammar as closely as the generator here, so testing them against each other gives me some confidence. Although I only know the parser is not too restrictive, but it could still be too permissive. :thinking: